motionplan/tpSpaceRRT.go

```diff
@@ -845,16 +846,86 @@ func (mp *tpSpaceRRTMotionPlanner) make2DTPSpaceDistanceOptions(ptg tpspace.PTGS
+func generateHeuristic(firstEdge, pathLen int) []float64 {
+	// The heuristic implemented here takes in the firstEdge and defines a lookAhead.
+	// For nodes in each direction around firstEdge, the algorithm will increment by one
+	// until it reaches firstEdge + lookAhead and firstEdge - lookAhead indices
+	// From then on, it will decrement by one until it gets to the edge of the list
+	// All values are passed into a softmax to convert the real-valued heuristics into a probability distribution
+
+	// It is better to have a lookAhead that is shorter because biasing towards collapsing shorter paths is
+	// more likely to yield success than finding one long one. Another observation to note is that sampling
+	// edges that are not connectable is very costly because the algorithm cycles through all PTGs in hopes
+	// of connecting them. Thus, finding short, connectable paths is best
+	lookAhead := 3.0
+	heuristics := make([]float64, pathLen)
+	for i := 0; i < pathLen; i++ {
+		// Creates ascending list until lookAhead and then descending list
+		heuristics[i] = math.Pow(math.Max(1, lookAhead-math.Abs(lookAhead-math.Abs(float64(firstEdge-i)))), 2)
+	}
 
-	maxCost := math.Inf(-1)
-	for _, wp := range path {
-		if wp.Cost() > maxCost {
-			maxCost = wp.Cost()
+	// firstEdge + adjacent edges should not be sampled, so set their heuristic to -Inf
+	// This gets converted to zero probability by softmax
+	heuristics[firstEdge] = math.Inf(-1)
+	if firstEdge-1 >= 0 {
+		heuristics[firstEdge-1] = math.Inf(-1)
+	}
+	if firstEdge+1 < pathLen {
+		heuristics[firstEdge+1] = math.Inf(-1)
+	}
+	return heuristics
+}
+
+// softmax takes in a heuristic list and converts it into a proability distribution function
+// This means the sum of the returned softmaxArr is equal to one.
+func softmax(heuristics []float64) []float64 {
+	sum := 0.0
+	for _, heuristic := range heuristics {
+		sum += math.Exp(heuristic)
+	}
+
+	softmaxArr := make([]float64, len(heuristics))
+	// Account for case where sum equals zero. In this case, assign equal weight to all indices
+	if sum == 0 {
+		for i := 0; i < len(heuristics); i++ {
+			softmaxArr[i] = 1.0 / float64(len(heuristics))
+		}
+	} else {
+		for i, heuristic := range heuristics {
+			softmaxArr[i] = math.Exp(heuristic) / sum
 		}
 	}
+	return softmaxArr
+}
+
+// generateCDF returns a cumulative distribution function that can be used to
+// sample the secondEdge for the smoothing algorithm.
+func generateCDF(firstEdge, pathLen int) []float64 {
+	heuristics := generateHeuristic(firstEdge, pathLen)
+	softmaxArr := softmax(heuristics)
+
+	// sum all successive values in the softmaxArr to convert it from a PDF to a CDF
+	for i := 1; i < len(softmaxArr); i++ {
+		softmaxArr[i] += softmaxArr[i-1]
+	}
+	return softmaxArr
+}
+
+// binarySearch finds the first instance of an element in the CDF that is greater than the sampled value
+// This is the index of the bucket that is returned.
+func binarySearch(cdf []float64, sample float64) int {
```

Comment:
I don't see this function having utility outside of the case you've created.  I think we can lump this in with the `generateCDF` function and rename it to something like sample CDF.